src/compiler/checker.ts

```diff
@@ -5236,7 +5236,7 @@ namespace ts {
 
         function createTupleType(elementTypes: Type[], thisType?: Type) {
             const id = getTypeListId(elementTypes) + "," + (thisType ? thisType.id : 0);
-            return tupleTypes[id] || (tupleTypes[id] = createNewTupleType(elementTypes, thisType));
+            return getOrUpdate(tupleTypes, id, () => createNewTupleType(elementTypes, thisType));

```

Comment:
This introduces a new closure and environment record for every call to `getTypeFromTypeAliasReference`. The original approach does not.